File: OldProblem/2230/E.cpp
#include <bits/stdc++.h>
using namespace std;

struct FastScanner {
    static constexpr int BUFFER_SIZE = 1 << 20;
    int idx = 0, size = 0;
    char buffer[BUFFER_SIZE];

    inline char get_char() {
        if (idx >= size) {
            size = (int)fread(buffer, 1, BUFFER_SIZE, stdin);
            idx = 0;
            if (size == 0) return 0;
        }
        return buffer[idx++];
    }

    int next_int() {
        char ch = get_char();
        while (ch && (ch < '0' || ch > '9')) ch = get_char();

        int value = 0;
        while (ch >= '0' && ch <= '9') {
            value = value * 10 + (ch - '0');
            ch = get_char();
        }
        return value;
    }
};

int main() {
    FastScanner fs;

    int n = fs.next_int();
    vector<int> p_values(n);
    for (int &x : p_values) x = fs.next_int();

    vector<pair<int, int>> points(n);
    for (int i = 0; i < n; ++i) {
        points[i] = {p_values[i], fs.next_int()};
    }

    sort(points.begin(), points.end());

    const int INF = 1'000'000'000;
    vector<int> p_front;
    vector<int> c_front;
    p_front.reserve(n);
    c_front.reserve(n);

    int best_c = INF;
    for (auto [p, c] : points) {
        if (c < best_c) {
            p_front.push_back(p);
            c_front.push_back(c);
            best_c = c;
        }
    }

    int k = (int)p_front.size();
    vector<int> logs(k + 1);
    for (int i = 2; i <= k; ++i) logs[i] = logs[i >> 1] + 1;

    int levels = logs[k] + 1;
    vector<vector<int>> sparse(levels, vector<int>(k));
    for (int i = 0; i < k; ++i) sparse[0][i] = p_front[i] + c_front[i];

    for (int level = 1; level < levels; ++level) {
        int len = 1 << level;
        int half = len >> 1;
        int usable = k - len + 1;
        for (int i = 0; i < usable; ++i) {
            sparse[level][i] = min(sparse[level - 1][i], sparse[level - 1][i + half]);
        }
    }

    vector<int> neg_c(k);
    for (int i = 0; i < k; ++i) neg_c[i] = -c_front[i];

    auto min_sum = [&](int left, int right) -> int {
        if (left >= right) return INF;
        int level = logs[right - left];
        int width = 1 << level;
        return min(sparse[level][left], sparse[level][right - width]);
    };

    auto count_c_ge = [&](int value) -> int {
        return (int)(upper_bound(neg_c.begin(), neg_c.end(), -value) - neg_c.begin());
    };

    auto has_common = [](int left_a, int right_a, int left_b, int right_b) -> bool {
        return max(left_a, left_b) < min(right_a, right_b);
    };

    auto best_const = [&](int left_a, int right_a, int left_b, int right_b, int value) -> int {
        return has_common(left_a, right_a, left_b, right_b) ? value : INF;
    };

    auto best_by_p = [&](int left_a, int right_a, int left_b, int right_b, int extra) -> int {
        int left = max(left_a, left_b);
        int right = min(right_a, right_b);
        return left < right ? p_front[left] + extra : INF;
    };

    auto best_by_c = [&](int left_a, int right_a, int left_b, int right_b, int extra) -> int {
        int left = max(left_a, left_b);
        int right = min(right_a, right_b);
        return left < right ? c_front[right - 1] + extra : INF;
    };

    auto best_by_sum = [&](int left_a, int right_a, int left_b, int right_b) -> int {
        int left = max(left_a, left_b);
        int right = min(right_a, right_b);
        return left < right ? min_sum(left, right) : INF;
    };

    auto solve_user = [&](int tp, int tc, int d) -> int {
        int p_mid = (int)(lower_bound(p_front.begin(), p_front.end(), tp) - p_front.begin());
        int p_full = (int)(lower_bound(p_front.begin(), p_front.end(), tp + d) - p_front.begin());

        int c_full = count_c_ge(tc + d);
        int c_mid = count_c_ge(tc);

        if (has_common(0, p_mid, c_mid, k)) return 0;

        int ans = INF;
        ans = min(ans, best_by_p(p_mid, p_full, c_mid, k, 0));
        ans = min(ans, best_const(p_full, k, c_mid, k, tp + d));
        ans = min(ans, best_by_c(0, p_mid, c_full, c_mid, 0));
        ans = min(ans, best_const(0, p_mid, 0, c_full, tc + d));
        ans = min(ans, best_by_sum(p_mid, p_full, c_full, c_mid));
        ans = min(ans, best_by_p(p_mid, p_full, 0, c_full, tc + d));
        ans = min(ans, best_by_c(p_full, k, c_full, c_mid, tp + d));
        ans = min(ans, best_const(p_full, k, 0, c_full, tp + tc + 2 * d));
        return ans;
    };

    int m = fs.next_int();
    vector<int> tp_values(m), tc_values(m);
    for (int &x : tp_values) x = fs.next_int();
    for (int &x : tc_values) x = fs.next_int();

    string output;
    output.reserve((size_t)m * 4);
    for (int i = 0; i < m; ++i) {
        int d = fs.next_int();
        output += to_string(solve_user(tp_values[i], tc_values[i], d));
        output += '\n';
    }

    fwrite(output.data(), 1, output.size(), stdout);
    return 0;
}

File: OldProblem/2230/E.py
import sys
from bisect import bisect_left, bisect_right


INF = 10**30


def debug(*x, **y): print(*x, file=sys.stderr, **y)


def read_ints():
    data = sys.stdin.buffer.read()
    value = 0
    in_number = False
    for byte in data:
        if 48 <= byte <= 57:
            value = value * 10 + byte - 48
            in_number = True
        elif in_number:
            yield value
            value = 0
            in_number = False
    if in_number:
        yield value


def run():
    ints = read_ints()
    n = next(ints)
    p_values = [next(ints) for _ in range(n)]
    points = [(p_values[i], next(ints)) for i in range(n)]
    del p_values

    points.sort()
    # Keep only Pareto-minimal articles.  If another article has both
    # coordinates no larger, it is never worse for any monotone clamp query.
    p_front = []
    c_front = []
    best_c = INF
    for p, c in points:
        if c < best_c:
            p_front.append(p)
            c_front.append(c)
            best_c = c

    def build_sparse_table(values):
        n = len(values)
        logs = [0] * (n + 1)
        for i in range(2, n + 1):
            logs[i] = logs[i >> 1] + 1
        table = [values]
        level = 1
        while (1 << level) <= n:
            half = 1 << (level - 1)
            prev = table[-1]
            length = n - (1 << level) + 1
            table.append([min(prev[i], prev[i + half]) for i in range(length)])
            level += 1
        return logs, table

    k = len(p_front)
    sum_front = [p_front[i] + c_front[i] for i in range(k)]
    logs, sparse = build_sparse_table(sum_front)
    neg_c_front = [-c for c in c_front]

    def min_sum(left, right):
        if left >= right:
            return INF
        level = logs[right - left]
        width = 1 << level
        row = sparse[level]
        return min(row[left], row[right - width])

    def count_c_ge(value):
        # c_front is strictly decreasing, so -c_front is increasing.
        return bisect_right(neg_c_front, -value)

    def has_common(left_a, right_a, left_b, right_b):
        return max(left_a, left_b) < min(right_a, right_b)

    def best_const(left_a, right_a, left_b, right_b, value):
        if has_common(left_a, right_a, left_b, right_b):
            return value
        return INF

    def best_by_p(left_a, right_a, left_b, right_b, extra):
        left = max(left_a, left_b)
        right = min(right_a, right_b)
        if left < right:
            return p_front[left] + extra
        return INF

    def best_by_c(left_a, right_a, left_b, right_b, extra):
        left = max(left_a, left_b)
        right = min(right_a, right_b)
        if left < right:
            return c_front[right - 1] + extra
        return INF

    def best_by_sum(left_a, right_a, left_b, right_b):
        left = max(left_a, left_b)
        right = min(right_a, right_b)
        if left < right:
            return min_sum(left, right)
        return INF

    def solve_user(tp, tc, d):
        # p zones on the frontier:
        #   [0, p_mid)      has political cost 0
        #   [p_mid, p_full) has political cost p
        #   [p_full, k)     has political cost tp + d
        p_mid = bisect_left(p_front, tp)
        p_full = bisect_left(p_front, tp + d)

        # c_front is decreasing, so high culture values are on the left:
        #   [0, c_full)     has cultural cost tc + d
        #   [c_full, c_mid) has cultural cost c
        #   [c_mid, k)      has cultural cost 0
        c_full = count_c_ge(tc + d)
        c_mid = count_c_ge(tc)

        if has_common(0, p_mid, c_mid, k):
            return 0

        return min(
            best_by_p(p_mid, p_full, c_mid, k, 0),
            best_const(p_full, k, c_mid, k, tp + d),
            best_by_c(0, p_mid, c_full, c_mid, 0),
            best_const(0, p_mid, 0, c_full, tc + d),
            best_by_sum(p_mid, p_full, c_full, c_mid),
            best_by_p(p_mid, p_full, 0, c_full, tc + d),
            best_by_c(p_full, k, c_full, c_mid, tp + d),
            best_const(p_full, k, 0, c_full, tp + tc + 2 * d),
        )

    m = next(ints)
    tp_values = [next(ints) for _ in range(m)]
    tc_values = [next(ints) for _ in range(m)]

    out = []
    for i in range(m):
        tp = tp_values[i]
        tc = tc_values[i]
        d = next(ints)
        out.append(str(solve_user(tp, tc, d)))

    return "\n".join(out)


if __name__ == "__main__":
    output = run()
    if output is not None:
        sys.stdout.write(str(output))
